### ISSUE TITLE
Parallel compilation fails on high core count

### ISSUE THREAD

**Issue by tkarna** [firedrakeproject/firedrake#1145]
I get compilation errors on our cluster if I use high enough core count.

```
Traceback (most recent call last):
  File "baroclinic_eddies.py", line 269, in <module>
    parse_options()
  File "baroclinic_eddies.py", line 265, in parse_options
    run_problem(**args_dict)
  File "baroclinic_eddies.py", line 156, in run_problem
    solver_obj.create_equations()
  File "/home/karnat/sources/firedrake/firedrake/src/thetis/thetis/solver.py", line 863, in create_equations
    self.fields.max_h_diff.assign(max_diff_alpha/self.dt * self.fields.h_elem_size_3d**2)
  File "<decorator-gen-272>", line 2, in assign
  File "/home/karnat/sources/firedrake/firedrake/src/firedrake/firedrake/utils.py", line 61, in wrapper
    return f(*args, **kwargs)
  File "/home/karnat/sources/firedrake/firedrake/src/firedrake/firedrake/function.py", line 362, in assign
    assemble_expressions.Assign(self, expr), subset)
  File "<decorator-gen-277>", line 2, in evaluate_expression
  File "/home/karnat/sources/firedrake/firedrake/src/firedrake/firedrake/utils.py", line 58, in wrapper
    return f(*args, **kwargs)
  File "/home/karnat/sources/firedrake/firedrake/src/firedrake/firedrake/assemble_expressions.py", line 575, in evaluate_expression
    evaluate_preprocessed_expression(k, args, subset)
  File "/home/karnat/sources/firedrake/firedrake/src/firedrake/firedrake/assemble_expressions.py", line 543, in evaluate_preprocessed_expression
    op2.par_loop(kernel, itset, *parloop_args)
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/base.py", line 4404, in par_loop
    return _make_object('ParLoop', kernel, it_space, *args, **kwargs).enqueue()
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/base.py", line 100, in enqueue
    _trace.append(self)
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/base.py", line 119, in append
    computation._run()
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/base.py", line 4083, in _run
    return self.compute()
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/base.py", line 4124, in compute
    fun = self._jitmodule
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/utils.py", line 62, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/sequential.py", line 896, in _jitmodule
    pass_layer_arg=self._pass_layer_arg)
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/caching.py", line 199, in __new__
    obj = make_obj()
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/caching.py", line 189, in make_obj
    obj.__init__(*args, **kwargs)
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/sequential.py", line 725, in __init__
    self.compile()
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/sequential.py", line 811, in compile
    comm=self.comm)
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/compilation.py", line 385, in load
    dll = compiler.get_so(src, extension)
  File "/home/karnat/sources/firedrake/firedrake/src/PyOP2/pyop2/compilation.py", line 270, in get_so
    return ctypes.CDLL(soname)
  File "/home/karnat/sources/firedrake/lib/python3.6/ctypes/__init__.py", line 348, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: /home/karnat/sources/firedrake/firedrake/.cache/pyop2/e2fd07c5e1f8e06bab97b097aad89326.so: cannot open shared object file: No such file or directory
```

Running with a low number of processes (~10) works fine. Errors appear only with higher count (~80). In all cases the `.so` file exists. Once all the kernels have been compiled on a low number of processes, the simulation works with 80 processes as well.

Looks like a disk race condition. Indeed, I can verify that while rank 0 (and some other lower ranks) can see the `.so` file, the higher ranks do not see it yet (`os.path.isfile(soname) == False`).  It seems that it may take up to 30 s before all the ranks can see the file.

This is probably a filesystem dependent issue, our file system is NFS. Trying to move the cache directory to another file system did not help. I cannot reproduce this issue on smaller Ubuntu machines.

**Comment by tkarna** [comment]
The proposed PyOP2 patch does fix this issue in my case.

**Comment by wence-** [comment]
Fixed via OP2/PyOP2#539 I hope.
